File: pos/views/terminal.py
"""POS sales screen (BR-013 .. BR-021, use case 2).

Workflow: search products -> cart -> enter customer number -> the server
detects an existing customer and applies the loyalty discount -> payment ->
invoice. The browser previews numbers; the server decides them.
"""

import uuid

from django.contrib import messages
from django.http import JsonResponse
from django.shortcuts import redirect, render
from django.views.decorators.http import require_GET, require_POST

from .. import customers as customer_service
from ..forms import CheckoutForm
from ..models import Product, StoreSetting
from ..permissions import POS_SELL, require
from ..services import BusinessRuleError, create_sale, price_cart

TXN_TOKEN_KEY = 'pos_txn_token'
COMPLETED_TOKENS_KEY = 'pos_completed_tokens'


def _issue_txn_token(request):
    """One-shot token that makes a double-submitted checkout idempotent."""
    token = uuid.uuid4().hex
    request.session[TXN_TOKEN_KEY] = token
    return token


@require(POS_SELL)
def pos_terminal(request):
    store = StoreSetting.load()
    products = Product.objects.active().select_related('category').order_by('name')
    term = request.GET.get('q', '').strip()
    if term:
        products = products.search(term)

    return render(
        request,
        'pos/terminal.html',
        {
            'page_title': 'New Sale',
            'products': products[:60],
            'search_term': term,
            'tax_rate': store.tax_rate,
            'loyalty_percent': store.existing_customer_discount_percent,
            'txn_token': _issue_txn_token(request),
            'payment_methods': [
                ('CASH', 'Cash'), ('CARD', 'Card'), ('MOBILE', 'Mobile'), ('BANK', 'Bank'),
            ],
        },
    )


@require(POS_SELL)
@require_GET
def product_lookup(request):
    """Type-ahead / barcode lookup for the sales screen (BR-005)."""
    term = request.GET.get('q', '').strip()
    queryset = Product.objects.active().select_related('category')
    if term:
        queryset = queryset.search(term)
    results = [
        {
            'id': p.id,
            'name': p.name,
            'sku': p.sku,
            'barcode': p.barcode or '',
            'category': p.category.name,
            'price': str(p.selling_price),
            'discount_percent': str(p.discount_percent),
            'final_price': str(p.final_price),
            'image_url': p.image_url,
            'initials': p.initials,
            'stock': p.stock_quantity,
            'unit': p.get_unit_display(),
            'status': p.stock_status,
        }
        for p in queryset.order_by('name')[:40]
    ]
    return JsonResponse({'results': results})


@require(POS_SELL)
@require_GET
def customer_lookup(request):
    """Tell the till whether this number is an existing customer.

    Informational only -- :func:`~pos.services.create_sale` decides again from
    the database when the sale is submitted.
    """
    result = customer_service.lookup(request.GET.get('number', ''))
    return JsonResponse(
        {
            'number': result['number'],
            'valid': result['valid'],
            'existing': result['existing'],
            'orders': result['orders'],
            'discount_percent': str(result['discount_percent']),
        }
    )


@require(POS_SELL)
@require_GET
def cart_quote(request):
    """Authoritative totals for the current cart -- the browser never decides money."""
    raw = request.GET.get('cart', '')
    pairs = []
    for chunk in raw.split(','):
        if not chunk.strip():
            continue
        try:
            product_id, quantity = chunk.split(':')
            pairs.append((int(product_id), int(quantity)))
        except ValueError:
            return JsonResponse({'error': 'Malformed cart.'}, status=400)

    store = StoreSetting.load()
    number = customer_service.normalise_customer_number(request.GET.get('customer_number', ''))
    existing = customer_service.is_existing_customer(number) if number else False
    loyalty = store.existing_customer_discount_percent if existing else '0.00'

    if not pairs:
        zero = '0.00'
        return JsonResponse(
            {
                'subtotal': zero, 'product_discount': zero, 'customer_discount': zero,
                'discount': zero, 'tax': zero, 'total': zero,
                'existing_customer': existing,
                'customer_discount_percent': str(loyalty),
                'lines': [],
            }
        )

    products = {p.pk: p for p in Product.objects.active().filter(pk__in=[p for p, _ in pairs])}
    items = [(products[pid], qty) for pid, qty in pairs if pid in products]
    if not items:
        return JsonResponse({'error': 'No valid products in the cart.'}, status=400)

    try:
        totals = price_cart(items, existing_customer=existing, store=store)
    except (BusinessRuleError, ArithmeticError) as exc:
        return JsonResponse({'error': str(exc)}, status=400)

    return JsonResponse(
        {
            'subtotal': str(totals.subtotal),
            'product_discount': str(totals.product_discount),
            'customer_discount': str(totals.customer_discount),
            'customer_discount_percent': str(totals.customer_discount_percent),
            'discount': str(totals.discount_amount),
            'net_before_tax': str(totals.net_before_tax),
            'tax': str(totals.tax_amount),
            'tax_rate': str(totals.tax_rate),
            'total': str(totals.total),
            'existing_customer': totals.existing_customer,
            'lines': [
                {
                    'id': line.product.id,
                    'name': line.product.name,
                    'quantity': line.quantity,
                    'unit_price': str(line.unit_price),
                    'discount_percent': str(line.discount_percent),
                    'discount': str(line.discount_amount),
                    'subtotal': str(line.subtotal),
                    'stock': line.product.stock_quantity,
                }
                for line in totals.lines
            ],
        }
    )


@require(POS_SELL)
@require_POST
def checkout(request):
    """Complete the sale (BR-021) and hand the cashier the receipt."""
    form = CheckoutForm(request.POST)
    if not form.is_valid():
        for error in form.errors.values():
            messages.error(request, error[0])
        return redirect('pos:pos_terminal')

    data = form.cleaned_data
    token = data['txn_token']
    completed = request.session.get(COMPLETED_TOKENS_KEY, {})

    # A refresh or double-click replays the same token: show the original sale
    # rather than charging the customer twice.
    if token in completed:
        messages.info(request, 'That sale was already completed.')
        return redirect('pos:invoice', pk=completed[token])
    if request.session.get(TXN_TOKEN_KEY) != token:
        messages.error(request, 'This sale has expired. Please rebuild the cart.')
        return redirect('pos:pos_terminal')

    try:
        sale = create_sale(
            cashier=request.user,
            items=data['cart'],
            customer_number=data.get('customer_number', ''),
            payment_method=data['payment_method'],
            amount_paid=data.get('amount_paid'),
            note=data.get('note', ''),
            request=request,
        )
    except BusinessRuleError as exc:
        messages.error(request, str(exc))
        return redirect('pos:pos_terminal')

    completed[token] = sale.pk
    # Keep the map small; only recent tokens can realistically be replayed.
    request.session[COMPLETED_TOKENS_KEY] = dict(list(completed.items())[-10:])
    request.session.pop(TXN_TOKEN_KEY, None)

    messages.success(request, f'Sale {sale.invoice_no} completed.')
    return redirect('pos:invoice', pk=sale.pk)

File: pos/static/pos/js/terminal.js
/* POS terminal: search, cart, customer lookup and totals.

   Money is never decided here. The browser previews nothing it invents: the
   server returns the authoritative quote for the cart, and decides customer
   eligibility again when the sale is submitted. This file only moves the
   cashier through search -> cart -> payment as quickly as possible. */
(function () {
  'use strict';

  var config = window.POS_CONFIG || {};
  var cart = new Map();          // productId -> {id, name, price, discount, final, stock, qty}
  var results = [];              // what is currently on screen, for keyboard use
  var cursor = -1;               // highlighted result index
  var paymentMethod = 'CASH';
  var quoteTimer = null;
  var searchTimer = null;
  var customerTimer = null;
  var currentTotal = 0;

  var els = {
    search: document.getElementById('product-search'),
    resultList: document.getElementById('search-results'),
    resultEmpty: document.getElementById('search-empty'),
    resultCount: document.getElementById('result-count'),
    status: document.getElementById('search-status'),
    lines: document.getElementById('cart-lines'),
    cartCount: document.getElementById('cart-count'),
    cartField: document.getElementById('cart-field'),
    subtotal: document.getElementById('t-subtotal'),
    productDiscount: document.getElementById('t-product-discount'),
    customerDiscount: document.getElementById('t-customer-discount'),
    customerLabel: document.getElementById('t-customer-label'),
    tax: document.getElementById('t-tax'),
    total: document.getElementById('t-total'),
    submit: document.getElementById('complete-sale'),
    clear: document.getElementById('clear-cart'),
    methodField: document.getElementById('payment-method'),
    cashBlock: document.getElementById('cash-block'),
    amountPaid: document.getElementById('amount-paid'),
    changeNote: document.getElementById('change-note'),
    customerNumber: document.getElementById('customer-number'),
    customerStatus: document.getElementById('customer-status'),
    form: document.getElementById('checkout-form')
  };

  function money(value) {
    return config.currency + ' ' + Number(value).toFixed(2);
  }

  function icon(name, cls) {
    var svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    svg.setAttribute('class', 'icon ' + (cls || ''));
    svg.setAttribute('aria-hidden', 'true');
    var use = document.createElementNS('http://www.w3.org/2000/svg', 'use');
    use.setAttribute('href', '#i-' + name);
    svg.appendChild(use);
    return svg;
  }

  function say(message, tone) {
    if (!els.status) { return; }
    els.status.textContent = message || '';
    els.status.className = 'pos-search__status ' + (tone === 'error' ? 'small' : 'muted small');
    if (tone === 'error') { els.status.style.color = 'var(--danger-dark)'; }
    else { els.status.style.color = ''; }
  }

  function serialiseCart() {
    var parts = [];
    cart.forEach(function (line) { parts.push(line.id + ':' + line.qty); });
    return parts.join(',');
  }

  // ------------------------------------------------------------------ cart
  function addToCart(product) {
    var existing = cart.get(product.id);
    var nextQty = existing ? existing.qty + 1 : 1;
    if (product.stock > 0 && nextQty > product.stock) {
      say(product.name + ': only ' + product.stock + ' in stock.', 'error');
      return;
    }
    if (existing) { existing.qty = nextQty; }
    else { cart.set(product.id, Object.assign({}, product, { qty: 1 })); }
    say(product.name + ' added to the cart.');
    renderCart();
  }

  function setQuantity(id, qty) {
    var line = cart.get(id);
    if (!line) { return; }
    if (qty < 1) { cart.delete(id); renderCart(); return; }
    if (line.stock > 0 && qty > line.stock) {
      say(line.name + ': only ' + line.stock + ' in stock.', 'error');
      renderCart();
      return;
    }
    line.qty = qty;
    renderCart();
  }

  function thumb(line) {
    if (line.image) {
      var img = document.createElement('img');
      img.className = 'cart-line__thumb';
      img.src = line.image;
      img.alt = '';
      img.loading = 'lazy';
      return img;
    }
    var span = document.createElement('span');
    span.className = 'cart-line__thumb';
    span.textContent = (line.name || '?').trim().charAt(0).toUpperCase();
    return span;
  }

  function qtyControl(line) {
    var box = document.createElement('div');
    box.className = 'qty';

    var minus = document.createElement('button');
    minus.type = 'button';
    minus.className = 'qty__btn';
    minus.setAttribute('aria-label', 'Decrease ' + line.name);
    minus.appendChild(icon('minus', 'icon--sm'));
    minus.addEventListener('click', function () { setQuantity(line.id, line.qty - 1); });

    var input = document.createElement('input');
    input.type = 'number';
    input.className = 'qty__input';
    input.min = '1';
    input.value = line.qty;
    input.setAttribute('aria-label', 'Quantity of ' + line.name);
    input.addEventListener('change', function () {
      setQuantity(line.id, parseInt(input.value, 10) || 0);
    });

    var plus = document.createElement('button');
    plus.type = 'button';
    plus.className = 'qty__btn';
    plus.setAttribute('aria-label', 'Increase ' + line.name);
    plus.appendChild(icon('plus', 'icon--sm'));
    plus.addEventListener('click', function () { setQuantity(line.id, line.qty + 1); });

    box.appendChild(minus);
    box.appendChild(input);
    box.appendChild(plus);
    return box;
  }

  function renderCart() {
    els.lines.textContent = '';
    var units = 0;

    if (cart.size === 0) {
      var note = document.createElement('div');
      note.className = 'empty';
      note.style.padding = '32px 20px';
      var strong = document.createElement('strong');
      strong.textContent = 'Cart is empty';
      var hint = document.createElement('div');
      hint.textContent = 'Search for a product to start the sale.';
      note.appendChild(strong);
      note.appendChild(hint);
      els.lines.appendChild(note);
    }

    cart.forEach(function (line) {
      units += line.qty;

      var row = document.createElement('div');
      row.className = 'cart-line';
      row.appendChild(thumb(line));

      var text = document.createElement('div');
      text.className = 'cart-line__text';

      var name = document.createElement('div');
      name.className = 'cart-line__name';
      name.title = line.name;
      name.textContent = line.name;
      text.appendChild(name);

      var price = document.createElement('div');
      price.className = 'cart-line__price';
      if (Number(line.discount) > 0) {
        price.appendChild(document.createTextNode(money(line.final) + ' '));
        var was = document.createElement('s');
        was.textContent = money(line.price);
        price.appendChild(was);
        price.appendChild(document.createTextNode(' · ' + Number(line.discount) + '% off'));
      } else {
        price.textContent = money(line.price) + ' each';
      }
      text.appendChild(price);
      row.appendChild(text);

      var side = document.createElement('div');
      side.className = 'cart-line__side';
      side.appendChild(qtyControl(line));
      var total = document.createElement('div');
      total.className = 'cart-line__total';
      total.textContent = money(Number(line.final) * line.qty);
      side.appendChild(total);
      row.appendChild(side);

      var remove = document.createElement('button');
      remove.type = 'button';
      remove.className = 'cart-line__remove';
      remove.setAttribute('aria-label', 'Remove ' + line.name);
      remove.appendChild(icon('trash', 'icon--sm'));
      remove.addEventListener('click', function () { setQuantity(line.id, 0); });
      row.appendChild(remove);

      els.lines.appendChild(row);
    });

    els.cartCount.textContent = units + ' item' + (units === 1 ? '' : 's');
    els.cartField.value = serialiseCart();
    els.submit.disabled = cart.size === 0;
    scheduleQuote();
  }

  // --------------------------------------------------------------- totals
  function scheduleQuote() {
    window.clearTimeout(quoteTimer);
    quoteTimer = window.setTimeout(requestQuote, 150);
  }

  function requestQuote() {
    if (cart.size === 0) {
      [els.subtotal, els.productDiscount, els.customerDiscount, els.tax, els.total]
        .forEach(function (el) { el.textContent = money(0); });
      currentTotal = 0;
      updateChange();
      return;
    }

    var params = new URLSearchParams({
      cart: serialiseCart(),
      customer_number: els.customerNumber.value || ''
    });

    fetch(config.quoteUrl + '?' + params.toString(), {
      headers: { 'X-Requested-With': 'XMLHttpRequest' }
    })
      .then(function (response) { return response.json(); })
      .then(function (data) {
        if (data.error) { say(data.error, 'error'); return; }
        els.subtotal.textContent = money(data.subtotal);
        els.productDiscount.textContent = '- ' + money(data.product_discount);
        els.customerDiscount.textContent = '- ' + money(data.customer_discount);
        els.customerLabel.textContent = Number(data.customer_discount_percent) > 0
          ? 'Customer Discount (' + Number(data.customer_discount_percent) + '%)'
          : 'Customer Discount';
        els.tax.textContent = money(data.tax);
        els.total.textContent = money(data.total);
        currentTotal = Number(data.total);
        updateChange();
      })
      .catch(function () { say('Could not reach the server to price the cart.', 'error'); });
  }

  function updateChange() {
    if (paymentMethod !== 'CASH') {
      els.changeNote.textContent = 'The exact total will be captured.';
      return;
    }
    var paid = Number(els.amountPaid.value || 0);
    if (!paid) {
      els.changeNote.textContent = 'Change is worked out once there is a total.';
    } else if (paid < currentTotal) {
      els.changeNote.textContent = 'Short by ' + money(currentTotal - paid);
    } else {
      els.changeNote.textContent = 'Change: ' + money(paid - currentTotal);
    }
  }

  // ------------------------------------------------------- customer lookup
  function setCustomerStatus(text, cls) {
    els.customerStatus.textContent = '';
    var span = document.createElement('span');
    span.className = cls;
    span.textContent = text;
    els.customerStatus.appendChild(span);
  }

  function checkCustomer() {
    var raw = (els.customerNumber.value || '').trim();
    if (!raw) {
      setCustomerStatus(
        'Optional. Enter it to check for a returning customer.', 'muted small'
      );
      scheduleQuote();
      return;
    }

    fetch(config.customerUrl + '?number=' + encodeURIComponent(raw), {
      headers: { 'X-Requested-With': 'XMLHttpRequest' }
    })
      .then(function (response) { return response.json(); })
      .then(function (data) {
        if (!data.valid) {
          setCustomerStatus(
            'Not a valid mobile number yet - for example 01712345678.', 'muted small'
          );
        } else if (data.existing) {
          setCustomerStatus(
            'Existing customer - ' + data.orders + ' previous order'
            + (data.orders === 1 ? '' : 's') + '. '
            + Number(data.discount_percent) + '% discount applied automatically.',
            'badge badge--success'
          );
        } else {
          setCustomerStatus('New customer - no customer discount.', 'badge');
        }
        scheduleQuote();
      })
      .catch(function () { say('Customer lookup is unavailable.', 'error'); });
  }

  // --------------------------------------------------------------- search
  function toProduct(item) {
    return {
      id: item.id,
      name: item.name,
      price: Number(item.price),
      discount: Number(item.discount_percent),
      final: Number(item.final_price),
      stock: item.stock,
      image: item.image_url || ''
    };
  }

  function highlight(index) {
    var rows = els.resultList.querySelectorAll('.result');
    if (!rows.length) { cursor = -1; return; }
    cursor = Math.max(0, Math.min(index, rows.length - 1));
    Array.prototype.forEach.call(rows, function (row, i) {
      row.classList.toggle('is-cursor', i === cursor);
      if (i === cursor && row.scrollIntoView) {
        row.scrollIntoView({ block: 'nearest' });
      }
    });
  }

  function renderResults(items) {
    results = items;
    cursor = -1;
    els.resultList.textContent = '';
    els.search.setAttribute('aria-expanded', items.length ? 'true' : 'false');

    if (!items.length) {
      els.resultEmpty.hidden = false;
      els.resultEmpty.innerHTML =
        '<div class="empty">'
        + '<div class="empty__icon"><svg class="icon icon--lg" aria-hidden="true">'
        + '<use href="#i-search"></use></svg></div>'
        + '<strong>No products match that search</strong>'
        + '<div>Check the spelling, or try the SKU instead.</div></div>';
      els.resultCount.textContent = '';
      return;
    }

    els.resultEmpty.hidden = true;
    els.resultCount.textContent = items.length + ' product' + (items.length === 1 ? '' : 's');

    items.forEach(function (item) {
      var row = document.createElement('button');
      row.type = 'button';
      row.className = 'result';
      row.setAttribute('role', 'option');
      if (item.stock <= 0) { row.disabled = true; }

      if (item.image_url) {
        var img = document.createElement('img');
        img.className = 'result__media';
        img.src = item.image_url;
        img.alt = '';
        img.loading = 'lazy';
        row.appendChild(img);
      } else {
        var blank = document.createElement('span');
        blank.className = 'result__media';
        blank.textContent = item.initials || '?';
        row.appendChild(blank);
      }

      var text = document.createElement('span');
      text.className = 'result__text';
      var name = document.createElement('span');
      name.className = 'result__name';
      name.textContent = item.name;
      var meta = document.createElement('span');
      meta.className = 'result__meta';
      meta.textContent = item.sku + ' · ' + item.category + ' · '
        + (item.stock <= 0
            ? 'Out of stock'
            : item.stock + ' ' + item.unit.toLowerCase() + ' in stock');
      text.appendChild(name);
      text.appendChild(meta);
      row.appendChild(text);

      var price = document.createElement('span');
      price.className = 'result__price';
      var strong = document.createElement('b');
      strong.textContent = money(item.final_price);
      price.appendChild(strong);
      if (Number(item.discount_percent) > 0) {
        price.appendChild(document.createElement('br'));
        var was = document.createElement('s');
        was.textContent = money(item.price);
        price.appendChild(was);
      }
      row.appendChild(price);

      row.addEventListener('click', function () {
        addToCart(toProduct(item));
        els.search.focus();
      });
      els.resultList.appendChild(row);
    });
  }

  function search(term) {
    term = (term || '').trim();
    if (term.length < 2) {
      results = [];
      els.resultList.textContent = '';
      els.resultCount.textContent = '';
      els.resultEmpty.hidden = false;
      els.resultEmpty.innerHTML =
        '<div class="empty">'
        + '<div class="empty__icon"><svg class="icon icon--lg" aria-hidden="true">'
        + '<use href="#i-search"></use></svg></div>'
        + '<strong>Nothing searched yet</strong>'
        + '<div>Type at least two characters, or scan a barcode to add it '
        + 'straight to the cart.</div></div>';
      return;
    }

    fetch(config.lookupUrl + '?q=' + encodeURIComponent(term), {
      headers: { 'X-Requested-With': 'XMLHttpRequest' }
    })
      .then(function (response) { return response.json(); })
      .then(function (data) {
        // A scanned barcode matches exactly one product: add it straight away
        // so the scanner never needs a second keystroke.
        if (data.results.length === 1 && data.results[0].barcode === term) {
          addToCart(toProduct(data.results[0]));
          els.search.value = '';
          els.search.select();
          renderResults([]);
          return;
        }
        renderResults(data.results);
      })
      .catch(function () { say('Product search is unavailable.', 'error'); });
  }

  // ----------------------------------------------------------------- wire
  els.search.addEventListener('input', function () {
    window.clearTimeout(searchTimer);
    var term = els.search.value;
    searchTimer = window.setTimeout(function () { search(term); }, 220);
  });

  els.search.addEventListener('keydown', function (event) {
    if (event.key === 'ArrowDown') {
      event.preventDefault();
      highlight(cursor + 1);
    } else if (event.key === 'ArrowUp') {
      event.preventDefault();
      highlight(cursor - 1);
    } else if (event.key === 'Enter') {
      event.preventDefault();
      window.clearTimeout(searchTimer);
      if (cursor >= 0 && results[cursor]) {
        if (results[cursor].stock > 0) { addToCart(toProduct(results[cursor])); }
        return;
      }
      // Nothing highlighted: run the search, which also handles a scan.
      search(els.search.value);
    } else if (event.key === 'Escape') {
      els.search.value = '';
      renderResults([]);
      search('');
    }
  });

  els.customerNumber.addEventListener('input', function () {
    window.clearTimeout(customerTimer);
    customerTimer = window.setTimeout(checkCustomer, 350);
  });
  els.customerNumber.addEventListener('blur', checkCustomer);

  Array.prototype.forEach.call(document.querySelectorAll('.pay-method'), function (button) {
    button.addEventListener('click', function () {
      Array.prototype.forEach.call(document.querySelectorAll('.pay-method'), function (other) {
        other.classList.remove('is-active');
        other.setAttribute('aria-pressed', 'false');
      });
      button.classList.add('is-active');
      button.setAttribute('aria-pressed', 'true');
      paymentMethod = button.dataset.method;
      els.methodField.value = paymentMethod;
      els.cashBlock.hidden = paymentMethod !== 'CASH';
      updateChange();
    });
  });

  els.amountPaid.addEventListener('input', updateChange);

  els.clear.addEventListener('click', function () {
    if (cart.size && !window.confirm('Clear the whole cart?')) { return; }
    cart.clear();
    els.customerNumber.value = '';
    checkCustomer();
    renderCart();
    els.search.focus();
  });

  els.form.addEventListener('submit', function (event) {
    if (cart.size === 0) {
      event.preventDefault();
      return;
    }
    if (paymentMethod === 'CASH') {
      var paid = Number(els.amountPaid.value || 0);
      if (paid < currentTotal) {
        event.preventDefault();
        say('Cash received is less than the total.', 'error');
        els.amountPaid.focus();
        return;
      }
    }
    // Guard against a double click resubmitting the same sale. The server has
    // its own one-shot token as well; this is only the friendly half.
    els.submit.disabled = true;
    els.submit.textContent = 'Processing…';
  });

  renderCart();
  if (els.search.value.trim()) { search(els.search.value); }
}());
